File: llang/L0/bin/main.ml

open Llang.Parser_utils
open Llang.Interpreter
open Llang.Typing
open Llang.Compiler_llvm

let () = print_endline "Welcome to the L language REPL & Compiler! (in OCaml)"

let () =
  print_endline "Insert an expression. Ctrl+D to exit.";
  let rec loop () =
    print_string "> "; flush stdout;
    match read_line () with
    | s ->
        let ast = parse_string s in
        begin try
          Printf.printf "\n-- Parser --\n";
          unparse_ast ast |> print_endline;
          (try
            Printf.printf "\n-- Untyped Interpreter --\n\n";
            ast |> eval |> string_of_value |> Printf.printf "= %s\n%!";
            with Failure msg ->
              Printf.eprintf "Interpreter Error: %s\n%!" msg);
          begin try
            Printf.printf "\n-- Type Checking --\n\n";
            let ast_typed = type_synth ast in
            ast_typed |> Llang.Ast_typed.type_of |> string_of_btype |> Printf.printf "= %s\n%!";
            (try
                Printf.printf "\n-- LLVM --\n\n";
                Llang.Llvm_utils.string_of_program ast_typed |> print_endline;
                Llang.Llvm_utils.write_llvm_text_file ast_typed "output.ll";
                Llang.Llvm_api_utils.write_bitcode_file ast_typed "output.bc"
              with Failure msg ->
                Printf.eprintf "LLVM Text Error: %s\n%!" msg);
            with Failure msg ->
              Printf.eprintf "Type Checking Error: %s\n%!" msg
          end;
          with Failure msg ->
            Printf.eprintf "Unparser Error: %s\n%!" msg
        end;
        loop ()
    | exception End_of_file -> print_endline "\nGoodbye!"
  in
  loop ()

File: llang/L0/rust/src/main.rs
use lalrpop_util::lalrpop_mod;
use std::io::{self, BufRead, Write};

use std::rc::Rc;
use environment::Env;
use crate::interpreter::Eval;
use crate::typing::TypeCheck;
use crate::codegen::Compile;

mod ast;
mod symbols;
mod interpreter;
mod environment;
mod ast_typed;
mod typing;
mod codegen;
mod llvm_utils;
mod llvm_generator;

lalrpop_mod!(pub grammar);
fn main() {
    let parser = grammar::CondParser::new();
    let stdin = io::stdin();
    let mut stdout = io::stdout();

    println!("Welcome to the L language REPL & Compiler! (in Rust)");

    loop {
        print!("> ");
        stdout.flush().unwrap();

        let mut line = String::new();
        if stdin.lock().read_line(&mut line).unwrap() == 0 {
            break;
        }

        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        if line == "exit" || line == "quit" {
            break;
        }

        match parser.parse(line) {
            Ok(expr) => 
                match expr.type_check(Rc::new(Env::new())) {
                    Ok(typed) => {
                        match expr.eval(Rc::new(Env::new())) {
                            Ok(val) => {
                                println!("{:?} = {:?}", expr, val);
                                match typed.compile() {
                                    Ok(cu) => {

                                        llvm_utils::print_compilation_unit(&cu);

                                        match llvm_generator::llvm_as_string(&cu) {
                                            Ok(ir) => println!("{}", ir),
                                            Err(err) => println!("failed to produce llvm code: {}", err),
                                        }

                                        match llvm_generator::write_llvm_text_file(&cu, "output.ll") {
                                            Ok(()) => println!("wrote LLVM IR to output.ll"),
                                            Err(err) => println!("failed to write output.ll: {}", err),
                                        }

                                        match llvm_generator::write_llvm_bitcode_to_file(&cu, "output.bc") {
                                            Ok(()) => println!("wrote LLVM IR to output.bc"),
                                            Err(err) => println!("failed to write output.bc: {}", err),
                                        }
                                    },
                                    Err(err) => println!("compile error: {}", err)
                                }
                            },
                            Err(err) => println!("Execution Error: {}", err)
                        }
                    },
                    Err(err) => println!("type error: {}", err)
                },
                Err(err) => println!("parse error: {}", err),
        }
    }
}


#[cfg(test)]
mod tests;
